src/threshold_monitor/threshold_monitor.1
.TH threshold_monitor.py 1 "$Date$"
.SH NAME
threshold_monitor.py \- compute peak-ground-acceleration (PGA) on data packets and issue threshold exceedance detections and alarms
.SH SYNOPSIS
.nf
\fBthreshold_monitor.py \fP[-s \fIstarttime\fP] [-e \fIendtime\fP]
                [-a \fIapi\fP] [-n \fInslc\fP] [-p \fIpfpath\fP]
                [-o \fIoutputdir\fP] [-v] [-l] [-b] 
.fi
.SH DESCRIPTION
\fBthreshold_monitor.py\fP is a program that continuously reads waveform packets from an Antelope orbserver,
Seedlink server, or synthesized packets from a DataScope CSS3.0 database, detrends, filters, and calibrates packets,
computes the PGA within each packet, compares PGA to pre-defined PGA thresholds for LOW, MEDIUM, and HIGH levels,
triggers threshold exceedance detections at the individual channel level, then groups detections at station level and
emits threshold alarms accordingly. It also tracks the latency of packets as they are retrieved, and emits latency alarms.
Packets for each station are processed in a separate thread so that problematic packets for one station have no impact on others.
Much of the heavy lifting is done by \fBdata_ingestion.py\fP, which handles all the parameter file reading, packet retrieval, buffering,
detrending, filtering, and latency monitoring. Waveform data from packets are merged with a longer data buffer to enable stable detrending
and filtering.

.SH OPTIONS
.IP "--api api or -a api"
Which API to use. This defaults to orb2obspy, which reads
waveform packets from an orbserver. Other choices are 
slink2obspy, which reads waveform packets from a Seedlink server,
and datascope2obspy, which reads simulated packets from a DataScope
database. These strings refer to separate programs that are 
imported by \fBdata_ingestion.py\fP called \fBorb2obspy.py\fP,
\fBslink2obspy.py\fP, and \fBdatascope2obspy.py\fP.
.IP "--starttime starttime or -s starttime"
Where to position the read pointer for the input
waveform packets. starttime must be a string understandable
by obspy.UTCDateTime(). This is only implemented for replaying
simulated waveform packets from a DataScope database, in
conjunction with the datascope2obspy API (see -a option) and an 
endtime (see -e option). If this option is not specified,
starttime defaults to the current date and time.
.IP "--endtime endtime or -e endtime"
When to stop reading waveform packets. endtime must be a string 
understandable by obspy.UTCDateTime(). This is useful not only
useful for replaying simulated waveform packets from a DataScope 
database, but is also useful for testing purposes, to stop the 
program at a certain time. However, the duration parameter in the
parameter file is usually used for the latter. If this option is 
not specified, endtime defaults to 2099/12/31, so the program will
run indefinitely.
.IP "-n nslc or --nslc nslc"
nslc is a string that is used to select waveform packets by SEED id. 
The general format is network.station.location.channel. It can
contain wildcards * and ?. Defaults to AK.*.*.HN? which selects
all strong motion channels on the AK network with a sampling rate of
>80 Hz. This includes all the Trans-Alaska Pipeline EMS strong motion
stations. This is only really used for testing purposes for 1 channel
or 1 station, because processing is already limited to the stations
specified in the thresholds array in the parameter file, which by default
restricts processing to Trans-Alaska Pipeline EMS stations.
.IP "-o outputdir or --outputdir outputdir"
Path to a directly for all outputs. This includes PNG files for latency
alarms, and threshold exceedance alarms, as well as CSV files that log
latency information, and PGA data, for each packet. There is one CSV file
per station.
.IP "-p parameterfile or --inputfile parameterfile"
Path to parameter file. This should be in YML format. 
This argument is optional and if not specified defaults to \fBthreshold_monitor.yml\fP.
A superset of command line parameters can be specified here, but command line parameters,
if given, will override the value specified in the parameter file.
.IP "-v or --verbose"
Verbose output flag. Useful for debugging.
.IP "-l or --latency"
Turns on latency tracking (and latency alarms).
.IP "-b or --benchmarking"
Turns on execution time tracking of different parts of the program, and outputs a summary
of this at the end of the program run. This is useful for code optimization, and uses
the timings class in \fBdata_ingestion.py\fP.

.SH "PROGRAM PARAMETER FILE"
The \fBthreshold_monitor.py\fP parameter file contains all of the information
necessary for specifying channels and filters for PGA computation and threshold monitoring.
It is specified in YML format. Following is an example \fBthreshold_monitor.py\fP parameter file.
.in 2c
.ft CW
.nf

.ne 5

#
#  threshold_monitor.py parameter file
#

.ne 12

# api to use
api: orb2obspy

# a datasource can be explicitly given
#datasource: 137.229.32.211:6520

# mode realtime or archive. archive is used to read old data, but setting a starttime in the past will set this anyway.
mode: realtime

# duration in seconds to run the program for. overwritten by command line variable endtime, if given. Default: -1 which means run forever.
duration: -1

# filter to use to process waveform data buffers. the length of the data buffer in seconds will be at least twice the period
# of the lowest filter frequency (e.g. 20-s for a 0.1 Hz high pass filter)
filterdef: 
  type: highpass
  freq: 
    - 0.1
  corners: 4

# default buffer length, in seconds. actual buffer length used will be whichever is greater of this, or that forced by filter specification above
bufferSecs: 10.0

# path to StationXML file
xmlfile: pipeline_stations.xml

# whether to just do a calibration correction, or perform a full instrument response removal
# since we are usually in the flat passband of a strong motion sensor, or a broadband seismometer, calibration correction usually suffices
remove_instrument_response: False

# number of seconds expected in a data packet. 1.0 for an orbserver. only really used in archive mode for chomping through a database, simulating packets of this size.
secondsPerPacket: 1.0

# net-sta-loc-chan or SEED id pattern to use. can be left as-is for Trans-Alaska Pipeline EMS stations as
# stations will only be used if they are defined in the thresholds dict AND they match this regular expression
nslc: AK.*..HN?  

# if latency of a data packet exceeds this (in seconds), send a latency alarm, and do not calculate PGA
maximum_latency: 600.0 

# block new threshold alarms at same station and state (e.g. LOW, MEDIUM, or HIGH) for this many seconds after an alarm
# higher state alarms will still go through (e.g. if we have a HIGH quickly following a MEDIUM, or MEDIUM or HIGH after a LOW)
threshold_alarm_timeout: 60.0 

# block new alarms at same station for this many seconds after a latency alarm
latency_alarm_timeout: 60.0 

# list of emails to send latency and threshold alarms to
email_list: 
- gthompson@alaska.edu
- thompsong@usf.edu

# thresholds. A station will only be used if defined here AND matches nslc filter regular expression.
# a separate thread is run for each station specied here, but not waveform packets would be collected if they
# did not also match the nslc regular expression
# each API has a select_stream() method that acts on this
thresholds: 
  PS01: 
    LOW: 0.05
    MEDIUM: 0.10
    HIGH: 0.15
  PS04: 
    LOW: 0.05
    MEDIUM: 0.10
    HIGH: 0.15
  PS11: 
    LOW: 0.05
    MEDIUM: 0.10
    HIGH: 0.15
  PS05: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.20
  PS06: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.20 
  PS07: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.20
  PS08: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.20
  PS09: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.20
  PS10: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.25
  PS12: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.25
  VMT: 
    LOW: 0.08
    MEDIUM: 0.15
    HIGH: 0.25

.
.SH AUTHOR
Glenn Thompson, gthompson@alaska.edu
.br
Alaska Earthquake Center
2024-08-14


src/threshold_monitor/threshold_monitor.py
#!/usr/bin/env python
from obspy import UTCDateTime
tstart = UTCDateTime()
import os
import sys
import numpy as np
import data_ingestion
import subprocess # for sending alarms
import pandas as pd
import matplotlib.pyplot as plt
import multiprocessing as mp
import re
import fcntl
mysql_imported = False
try:
    import mysql.connector as mysql
    mysql_imported = True
except:
    print('mysql not imported')
################################################################################
###                            CLASSES                                       ###
################################################################################

class thresholdHistory(object):
    ROWNUM = -1
    def __init__(self, thresholds, station, outputdir = None, seconds_to_keep=60):
        if outputdir: 
            self.outputdir = outputdir
        else:
            self.outputdir = UTCDateTime().isoformat()
        self.rownum = []
        self.seed_id = []
        self.starttime = []
        self.endtime = []
        self.peaktime = []
        self.value = []
        self.status = []
        self.thresholds = thresholds
        self.seconds_to_keep = seconds_to_keep
        self.last_trimmed_time = UTCDateTime()
        self.previous_state = {}
        self.secondsPerPacket = None
        self.outputdir = outputdir
        self.station = station
        self.csvfile = os.path.join(self.outputdir, f'threshold_history_{station}.csv')
        # start the output file
        row = 'rownum,seed_id,starttime,endtime,peaktime,value,status\n'
        data_ingestion.append_to_csvfile(self.csvfile, row) 

    def update(self, seed_id, starttime, endtime, peaktime, value, status):

        # update thresholdHistory attributes - these are all lists and there is one element per Trace from each packet Stream
        self.ROWNUM += 1
        self.rownum.append(self.ROWNUM)
        self.seed_id.append(seed_id)
        self.starttime.append(starttime)
        self.endtime.append(endtime)
        self.peaktime.append(peaktime)
        self.value.append(value)
        self.status.append(status)

        # update the output file
        row = f'{self.ROWNUM},{seed_id},{starttime},{endtime},{peaktime},{value},{status}' + '\n'
        data_ingestion.append_to_csvfile(self.csvfile, row)        
        
        # trim the object
        if UTCDateTime() > self.last_trimmed_time + self.seconds_to_keep:
            self.trim()
            data_ingestion.trim_csvfile(self.csvfile)
        
        # The following logic is designed to issue a threshold exceedance detection if the status changes upwards only, e.g. OFF -> LOW, or LOW -> MEDIUM, or MEDIUM -> HIGH
        thresholdDetection =  None
        if status == 'OFF' or not seed_id in self.previous_state: # Reset everything
            self.previous_state[seed_id] = {'status':status, 'value':value, 'peaktime':peaktime} 
        elif value>self.previous_state[seed_id]['value'] and status != self.previous_state[seed_id]['status']: # at least LOW, and value increased
            thresholdDetection = {'seed_id':seed_id, 'starttime':starttime, 'endtime':endtime, 'peaktime':peaktime, 'value':value, 'status':status}
        self.previous_state[seed_id] = {'status':status, 'value':value, 'peaktime':peaktime} 
        return thresholdDetection

    def to_dataframe(self):
        df = pd.DataFrame()
        df['rownum'] = self.rownum
        df['seed_id'] = self.seed_id
        df['starttime'] = self.starttime
        df['endtime'] = self.endtime
        df['peaktime'] = self.peaktime
        df['value'] = self.value
        df['status'] = self.status
        return df

    def print(self):
        df = self.to_dataframe()
        print(df)

    def plot(self, outfile='threshold_history.png', load_csv=False):
        timecol = 'starttime'
        if load_csv and self.ROWNUM > 0:
            df = pd.read_csv(self.csvfile)
            df['datetime'] = [UTCDateTime(tstr).datetime for tstr in df[timecol]]  
        else:
            self.trim() # trim so we always have a consistent 10-minute plot, or whatever seconds_to_keep is set to
            df = self.to_dataframe()
            df['datetime'] = [t.datetime for t in df[timecol]]
        units = 'm/s^2'
        seed_ids = df['seed_id'].unique()
        #print('seed_ids: ',seed_ids)
        fig, ax = plt.subplots()
        cols = ['black', 'blue', 'grey']
        for i, seed_id in enumerate(seed_ids):
            thisdf = df[df.seed_id == seed_id]
            ymax = thisdf['value'].max()
            ymin = thisdf['value'].min() 
            threshmax = max(self.thresholds[self.station].values())
            threshmin = min(self.thresholds[self.station].values())
            ymax = max((ymax, threshmax))
            ymin = min((ymin, threshmin))

            df2 = thisdf.copy()
            df2.reset_index(inplace=True)
            df2.plot(ax=ax, x='datetime', y='value', style='.-', label=seed_id[-1], \
                    title=f'peak amplitude vs. time for {self.station}?', 
                    ylim=[ymin/1.5, ymax*1.5], logy=True, color=cols[i])
        ax.set_xlabel(f"Date/Time on {df.loc[0, 'datetime'].strftime('%Y/%m/%d')}") # getting a string here, probably when loading from file
        ax.set_ylabel(f'Peak amplitude ({units})')
            
        handles, labels = ax.get_legend_handles_labels()
        cols = ['r', 'g', 'y']
        i = 0
        station_thresholds = self.thresholds[self.station]
        for k,v in station_thresholds.items():
            handles.append(ax.axhline(y=v, xmin=-1, xmax=1, color=cols[i], linestyle='--', lw=2, label=k.upper()))
            xlims = ax.get_xlim()
            ax.text((xlims[0]+xlims[1])/2, v, k.upper(), color=cols[i], fontsize=12)
            i+=1
        plt.legend(handles=handles)
        plt.savefig(outfile)
        plt.close()

    def trim(self):
        # trim the object lists
        # find index N where self.starttime is within the last self.seconds_to_keep seconds
        N = next(x for x, val in enumerate(self.starttime) if val>self.starttime[-1]-self.seconds_to_keep)
        self.rownum = self.rownum[N:]
        self.seed_id = self.seed_id[N:]
        self.starttime = self.starttime[N:]
        self.endtime = self.endtime[N:]
        self.peaktime = self.peaktime[N:]
        self.value = self.value[N:]
        self.status = self.status[N:]

################################################################################
class MyDataClient(data_ingestion.RealTimeDataClient):

    def __init__(self, params): 
        super().__init__(params)
        self.last_alarm = {'peaktime':UTCDateTime(1900,1,1), 
                           'status': 'OFF',
                           'value': 0.0
                           }

        g = 9.80665 # m/s^2
        for k, v in self.thresholds[self.station].items(): # convert thresholds from str and units g to units m/s**2
            self.thresholds[self.station][k] = float(v) * g
        self.thresholdHistoryObject = thresholdHistory(self.thresholds, self.station, outputdir=self.outputdir)
        if self.verbose:
            print('THRESHOLDS:')
            print(self.thresholds)
        
        # connect to mysql database
        mysql_info = params['mysql_info']
        if mysql_imported:
            self.db = mysql.connect(
                user = mysql_info['user'],
                password = mysql_info['password'],
                host = mysql_info['host'],
                database = mysql_info['database']
            )

    def computePGA(self):
        st = self.currentPacket
        pga_dict = dict()
        # find max absolute value, and time of that max value
        for tr in st:
            x = np.absolute(tr.data)
            x_max = np.max(x)
            ind_max = np.argmax(x)
            time_max = tr.stats.starttime + tr.stats.delta * ind_max
            # pga_dict has starttime and endtime of packet, peak value, and time of that peak value (which falls between start and end time)
            pga_dict[tr.id] = {'value':x_max, 'starttime':tr.stats.starttime, 'endtime':tr.stats.endtime, 'peaktime':time_max}
        return pga_dict

    def PGA2thresholddetections(self, tracemax):
        thresholdDetections = []
        station_thresholds = self.thresholds[self.station]
        for seed_id in tracemax.keys():
            this = tracemax[seed_id]
            status = 'OFF'

            # the logic here does not assume that 'OFF', 'LOW', 'MEDIUM' and 'HIGH' will always be the threshold labels we use
            # or that the threshold levels in station_thresholds are in numerically increasing order of threshold level
            # this allows us to change labels and/or levels in parameter file at any time
            highest_v = 0.0 # the highest threshold value triggered at
            for k, v in station_thresholds.items(): # at this point, k is like 'low' or 'medium', and v is corresponding threshold level in m/s^2
                if this['value'] > v and v > highest_v:
                    highest_v = v
                    status = k.upper() # e.g. turn 'low' into 'LOW'. 

            thisThresholdDetection = self.thresholdHistoryObject.update(seed_id, this['starttime'], this['endtime'], \
                                                                        this['peaktime'], this['value'], status)
            if thisThresholdDetection:
                thresholdDetections.append(thisThresholdDetection)
        return thresholdDetections            

    def send_alarm(self, seed_id, starttime, endtime, peaktime, value, status, thresholdDetections):
        now = UTCDateTime()
        subject = f"{status} threshold Alarm at {self.station} at {peaktime}"
        body = subject + '\n'
        for td in thresholdDetections:
            body += f"Threshold Alarm at {td['seed_id']} at {td['peaktime'].strftime('%Y-%m-%dT%H:%M:%S')} exceeded {td['status']} Threshold. Level now {td['value']}"
        pngfile = os.path.join(self.outputdir, f'threshold_alarm_{peaktime.strftime("%Y%m%d%H%M%S%F")}_{self.station}_{status}.png')
        self.thresholdHistoryObject.plot(outfile=pngfile, load_csv=False)
        data_ingestion.send_email_alarm(subject, body, self.email_list, pngfile=pngfile, verbose=True)

        # send update to mysql database
        if mysql_imported:
            query_cursor = self.db.cursor()
            if self.station == "VMT":
                sta_id = 13
            else:
                sta_id = int(self.station[-2:])

            if status in ["LOW", "MED", "HIGH"]:
                query = f'''UPDATE occ_display SET {status.lower()}=1 WHERE station_id={sta_id};'''
            else:
                query = f'''UPDATE occ_display SET 'low'=0, 'med'=0, 'high'=0 WHERE station_id={sta_id};'''
            
            query_cursor.execute(query)
            self.db.commit()


    def thresholddetections2alarms(self, thresholdDetections):
        ''' force alarm only at station level, not individual channels
        so we now track last alarm issued and use threshold_alarm_timeout parameter '''
        #print('thresholdDetections = ', thresholdDetections) # probably important enough to log even if verbose=False

        ''' Note that we have 1 station per thread, so we can only have 3 channels here. And we collapse from potentially
        3 threshold exceedance detections to 1 alarm based on the channel with the highest PGA value '''
        maxvalue = 0.0
        for td in thresholdDetections:
            if td['value'] > maxvalue:
                maxvalue = td['value']
                status = td['status']
                peaktime = td['peaktime']
                starttime = td['starttime']
                endtime = td['endtime']
                seed_id = td['seed_id']

        ''' We still only send an alarm if we are beyond the threshold_alarm_timeout period OR the status has increased, e.g. from LOW to MEDIUM'''
        if peaktime > self.last_alarm['peaktime'] + self.threshold_alarm_timeout or (maxvalue > self.last_alarm['value'] and status!=self.last_alarm['status']):
            self.send_alarm(seed_id, starttime, endtime, peaktime, maxvalue, status, thresholdDetections)
    
    def analyze(self):

        pga_dict = self.computePGA()
        self.update_timings('computing_max')

        # threshold exceedance
        thresholdDetections = self.PGA2thresholddetections(pga_dict)
        self.update_timings('threshold_exceedance')

        # alarm decision making
        if len(thresholdDetections) > 0:
            self.thresholddetections2alarms(thresholdDetections)

################################################################################
###                            FUNCTIONS                                     ###
################################################################################
def parse_station_matches(params): 
    ''' GT 20240811: this was added by Luke and it actually seems to require that station is defined in params['thresholds']
    so default is obsolete, and matches params['nslc'] which is a regular expression. not clear if this will
     work for antelope and non-antelope based data sources '''
    matched_nslc = []
    for k in params["thresholds"].keys():
        sta = params['nslc'].split('.')[1] # get just the station
        sta_regex = sta.replace('?', '.').replace('*', '.*') # switch wildcards for regex versions

        # check that this station matches the given nslc
        if re.fullmatch(sta_regex, k) is not None:
            # add to matched_nslc, replacing sta part with just this station
            sta_params = params.copy()
            sta_nslc = params["nslc"].split('.')
            sta_nslc[1] = k
            sta_params["nslc"] = '.'.join(sta_nslc)
            matched_nslc.append(sta_params)
    
    return matched_nslc

def run_parallel(params):
    datahandler = MyDataClient(params)
    datahandler.run()
    datahandler.close()
    return datahandler

def main(argv):

    ###########################################################################
    # SETTING UP PARAMETERS FROM COMMAND LINE & PARAMETER FILE
    ###########################################################################
    params = data_ingestion.get_params(argv) 
    
    ###########################################################################
    # THIS IS WHERE THE WORK GETS DONE
    ###########################################################################

    param_list = parse_station_matches(params)

    with mp.Pool(processes=len(param_list)) as mp_pool:
        datahandlers = mp_pool.map(run_parallel, param_list)
        mp_pool.close()
        mp_pool.join()

    ###########################################################################
    # THIS IS ALL ABOUT REPORTING WHAT HAPPENED
    ########################################################################### 

    for datahandler in datahandlers:
        datahandler.report()
        datahandler.thresholdHistoryObject.print() # unique
        datahandler.thresholdHistoryObject.plot(outfile=os.path.join(datahandler.outputdir, f'thresholds_{datahandler.station}.png'), load_csv=True) # seed_id is automatically added to the file pattern within plot function 

# THIS GETS CALLED WHEN PROGRAM RUN FROM COMMAND LINE        
if __name__ == "__main__":
    main(sys.argv) 
    t_end = UTCDateTime()
    print(f'Finished after {t_end - tstart} seconds')
